src/api/emscripten/mod.rs

```diff
@@ -380,18 +424,39 @@ fn em_try(res: ffi::EMSCRIPTEN_RESULT) -> Result<(), String> {
 extern fn mouse_callback(
         event_type: libc::c_int,
         event: *const ffi::EmscriptenMouseEvent,
-        event_queue: *mut libc::c_void) -> ffi::EM_BOOL {
+        user_data: *mut libc::c_void) -> ffi::EM_BOOL {
     unsafe {
         use std::mem;
-        let queue: &RefCell<VecDeque<Event>> = mem::transmute(event_queue);
+        use std::ptr;
+
+        let inner_window: &RefCell<InnerWindow> = mem::transmute(user_data);
+        let mut inner_window = inner_window.borrow_mut();
+
         match event_type {
             ffi::EMSCRIPTEN_EVENT_MOUSEMOVE => {
-                queue.borrow_mut().push_back(Event::MouseMoved(
-                        (*event).canvas_x as i32,
-                        (*event).canvas_y as i32));
+                // send relative move if cursor is grabbed
+                match inner_window.cursor_state {
+                    ::CursorState::Grab => inner_window.events.push_back(
+                        Event::MouseMoved(
+                            (*event).movement_x as i32,
+                            (*event).movement_y as i32)),
+                    _ => inner_window.events.push_back(
+                        Event::MouseMoved(
+                            (*event).canvas_x as i32,
+                            (*event).canvas_y as i32)),
+                }
             },
             ffi::EMSCRIPTEN_EVENT_MOUSEDOWN => {
-                queue.borrow_mut().push_back(Event::MouseInput(
+                // call pointerlock if needed
+                if let ::CursorState::Grab = inner_window.cursor_state {
+                    let mut pointerlock_status: ffi::EmscriptenPointerlockChangeEvent = mem::uninitialized();
```

Comment:
The weird thing is that get_pointerlock_status can fail, but I think initcontext can't fail? https://kripken.github.io/emscripten-site/docs/api_reference/html5.h.html#c.emscripten_get_pointerlock_status Anyways, the change looks good, thanks!